pkg/network/node/vxlan_monitor.go

```diff
@@ -0,0 +1,252 @@
+package node
+
+import (
+	"fmt"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/openshift/origin/pkg/util/ovs"
+)
+
+// egressVXLANMonitor monitors the health of automatic egress IPs by periodically checking
+// "ovs-ofctl dump-flows" output and noticing if the number of packets sent over VXLAN to
+// an egress node has increased, but the number of packets received over VXLAN from the
+// node hasn't. If that happens, it marks the node as being offline, and then starts
+// periodically pinging the node's SDN IP address, until the incoming packet count
+// increases, at which point it marks the node online again.
+//
+// The fact that we (normally) use pod-to-egress traffic to do the monitoring rather than
+// actively pinging the nodes means that if an egress node falls over while no one is
+// using an egress IP on it, we won't notice the problem until someone does try to use it.
+// So, eg, the first pod created in a namespace might spend 5 seconds trying to talk to a
+// dead egress node before falling over to its backup egress IP.
+type egressVXLANMonitor struct {
+	sync.Mutex
+
+	ovsif        ovs.Interface
+	updates      chan<- *egressVXLANNode
+	pollInterval time.Duration
+
+	monitorNodes map[string]*egressVXLANNode
+}
+
+type egressVXLANNode struct {
+	nodeIP  string
+	sdnIP   string
+	offline bool
+
+	in  uint64
+	out uint64
+
+	retries int
+}
+
+const (
+	defaultPollInterval = 5 * time.Second
+	repollInterval      = time.Second
+	maxRetries          = 2
+)
+
+func newEgressVXLANMonitor(ovsif ovs.Interface, updates chan<- *egressVXLANNode) *egressVXLANMonitor {
+	return &egressVXLANMonitor{
+		ovsif:        ovsif,
+		updates:      updates,
+		pollInterval: defaultPollInterval,
+		monitorNodes: make(map[string]*egressVXLANNode),
+	}
+}
+
+func (evm *egressVXLANMonitor) AddNode(nodeIP, sdnIP string) {
+	evm.Lock()
+	defer evm.Unlock()
+
+	if evm.monitorNodes[nodeIP] != nil {
+		return
+	}
+	glog.V(4).Infof("Monitoring node %s (SDN IP %s)", nodeIP, sdnIP)
+
+	evm.monitorNodes[nodeIP] = &egressVXLANNode{
+		nodeIP: nodeIP,
+		sdnIP:  sdnIP,
+	}
+	if len(evm.monitorNodes) == 1 && evm.pollInterval != 0 {
```

Comment:
Is it worth turning off polling in RemoveNode()?  Maybe not, just curious.

Reply:
Hm... I was doing it from poll() (it returned true if evm.monitorNodes was empty, causing the PollInfinite() to exit). But that creates a race condition if you delete the last node, then add a new node before poll() runs. So I redid it with a stop channel.